src/queue.jl

```diff
@@ -6,31 +6,57 @@ Each queue is uniquely associated with an agent.
 """
 mutable struct HSAQueue
     agent::HSAAgent
-    queue::Ref{Ptr{HSA.Queue}}
+    queue::Ptr{HSA.Queue}
     active::Bool
 end
-
+const QUEUES = WeakKeyDict{Ptr{HSA.Queue},HSAQueue}()
 const DEFAULT_QUEUES = IdDict{HSAAgent,HSAQueue}()
 
+function queue_error_handler(status::HSA.Status, _queue::Ptr{HSA.Queue}, data::Ptr{Cvoid})::Nothing
+    # TODO: From which thread are we calling this?
+    if status !== HSA.STATUS_SUCCESS then
+        queue = QUEUES[_queue]
+        queue.active = false
+        @assert DEFAULT_QUEUES[queue.agent] == queue # @jpsamaroo who gurantuees this 1:1 relation?
+        delete!(DEFAULT_QUEUES, queue.agent)
+        lock(RT_LOCK) do
+            delete!(_active_kernels, queue)
+        end
+        # do we need to call `HSA.queue_destroy`?
```

Comment:
I'm not particularly clear on the semantics, but I do know we're not allowed to destroy the queue in the queue callback, because `queue_destroy` waits on any running callbacks to terminate. I think we should probably `@async` destroy the queue.